Client for the IRIS Syngine Service

krischer opened the pull request:
This PR adds a new client `obspy.clients.syngine` to provide an interface with the [IRIS Syngine](http://service.iris.edu/irisws/syngine/1) service. It supports the full service including GET and POST requests and corresponding ObsPy methods.

Tests should be fairly comprehensive and are mix of mock and actual network tests that download stuff.

### Usage Examples

```python
>>> from obspy.clients.syngine import Client
>>> client = Client()

>>> st = client.get_waveforms(model="ak135f_5s", network="IU", station="ANMO",
                              eventid="GCMT:C201002270634A")
>>> print(st)
3 Trace(s) in Stream:
IU.ANMO.SE.MXZ | 2010-02-27T06:35:14... - ... | 4.0 Hz, 15520 samples
IU.ANMO.SE.MXN | 2010-02-27T06:35:14... - ... | 4.0 Hz, 15520 samples
IU.ANMO.SE.MXE | 2010-02-27T06:35:14... - ... | 4.0 Hz, 15520 samples
```

```python
>>> bulk = [{"latitude": 10.1, "longitude": 12.2, "stacode": "AA"},
...         {"latitude": 14.5, "longitude": 10.0, "stacode": "BB"}]
>>> st = client.get_waveforms_bulk(
...     model="ak135f_5s", eventid="GCMT:C201002270634A",
...     bulk=bulk, starttime="P-10", endtime="P+20")
>>> print(st)  
6 Trace(s) in Stream:
XX.AA.SE.MXZ | 2010-02-27T06:48:11.500000Z - ... | 4.0 Hz, 120 samples
XX.AA.SE.MXN | 2010-02-27T06:48:11.500000Z - ... | 4.0 Hz, 120 samples
XX.AA.SE.MXE | 2010-02-27T06:48:11.500000Z - ... | 4.0 Hz, 120 samples
XX.BB.SE.MXZ | 2010-02-27T06:48:15.250000Z - ... | 4.0 Hz, 120 samples
XX.BB.SE.MXN | 2010-02-27T06:48:15.250000Z - ... | 4.0 Hz, 120 samples
XX.BB.SE.MXE | 2010-02-27T06:48:15.250000Z - ... | 4.0 Hz, 120 samples
```

### Additional Things

* This PR introduces a new dependency to ObsPy: The [requests](http://docs.python-requests.org) module. It is one of the most used Python modules and thus should prove no problem to packaging people. I think we should base all future (and potentially) existing HTTP clients on the `requests` library - it just makes things much simpler.
* (@jkmacc-LANL) It also continues ideas from #1212 and introduces two new mix-in classes `RemoteBaseClient` (a base class for all remote clients) and `HTTPClient` (a base class for all clients downloading over HTTP). The later also contains actual downloading methods so it does not have to be reimplemented in each client. The Syngine client is based on these.

megies commented:
> a new client obspy.clients.syngine to provide an interface with the IRIS Syngine service

Maybe it's better placed under `clients/iris`? Or are there any plans to have such a server also at a different institution?

krischer commented:
> Or are there any plans to have such a server also at a different institution?

That was my thinking - I'm not sure if that ever happens but the client is general enough to be able to do it.

jkmacc-LANL commented:
Hot :smile: !  I suspect that it'd be hard to merge this with `iris.Client`, as they both have waveform methods.

QuLogic commented on CHANGELOG.txt:
```diff
@@ -46,6 +46,8 @@ master:
    * Removed the dedicated client. Data can still be accessed by using the FDSN
      client.
+ - obspy.clients.syngine:
+   * New client for the IRIS Syngine service.
```
A one-liner about what this service actually does would be good.

megies commented:
> Hot :smile: ! I suspect that it'd be hard to merge this with iris.Client, as they both have waveform methods.

Yeah, I just had the impression that the server-client connection was more IRIS-specific than instaseis/syngine specific. But I talked with @krischer and I don't mind if it stays like this.

megies commented:
This PR could also use a short info in https://github.com/obspy/obspy/blob/master/misc/docs/source/tutorial/code_snippets/retrieving_data_from_datacenters.rst.

CTrabant commented:
> Hot  ! I suspect that it'd be hard to merge this with iris.Client, as they both have waveform methods.
> 
> Yeah, I just had the impression that the server-client connection was more IRIS-specific than instaseis/syngine specific. But I talked with @krischer <https://github.com/krischer> and I don't mind if it stays like this.
> 
Well, for what it's worth, Syngine is very IRIS specific and is IRIS branding.  For most purposes it is no different than any other "irisws" services that IRIS maintains. That being said I would not be surprised to see other service implementations that are Syngine-compatible, similar to how others have implemented irisws-sacpz and irisws-resp; that will be great.

That being said I do not have a strong opinion on where the module lives and am very happy that a Syngine client will be in ObsPy, thanks @krishcer!

krischer commented:
Can this be merged? It is ready from my point of view.

megies commented:
> This PR could also use a short info in https://github.com/obspy/obspy/blob/master/misc/docs/source/tutorial/code_snippets/retrieving_data_from_datacenters.rst.

Would be good to have a short note here, otherwise good to merge I think.

QuLogic commented on obspy/clients/base.py:
```diff
@@ -50,31 +50,198 @@ def get_stations(self, network, station, location, channel, starttime,
+        :param data: If given, a POST request will be sent with the data in
+            the body of the request.
+        :type data: dictionary, bytes, or file-like object
+        :return: The response object resulting in the error.
```
This statement is the opposite of the second paragraph, which says it will only return if the code is 200, i.e., not an error.

In fact, I think whether it returns anything or not has nothing to do with the return code, but whether a location to save the result was specified.

QuLogic commented on obspy/clients/syngine/README.txt:
```diff
@@ -0,0 +1,26 @@
+Syngine service client for ObsPy.
+
+Contains requests abilities that naturally integrate with the rest of ObsPy
+for the IRIS syngine service (http://ds.iris.edu/ds/products/syngine/).
```
http->https

QuLogic commented on obspy/clients/syngine/README.txt:
```diff
@@ -0,0 +1,26 @@
+The goal of the ObsPy project is to facilitate rapid application development
+for seismology.
+
+For more information visit http://www.obspy.org.
```
http->https

QuLogic commented on obspy/clients/syngine/__init__.py:
```diff
@@ -0,0 +1,133 @@
+===========================================================
+
+This module offers methods to download from the IRIS syngine service
+(http://ds.iris.edu/ds/products/syngine/). The service is able to generate
```
http->https

QuLogic commented on obspy/clients/syngine/__init__.py:
```diff
@@ -0,0 +1,133 @@
+    The ObsPy Development Team (devs@obspy.org)
+:license:
+    GNU Lesser General Public License, Version 3
+    (http://www.gnu.org/copyleft/lesser.html)
```
http->https (This one redirects automatically so it saves a request, too!)

QuLogic commented on obspy/clients/syngine/__init__.py:
```diff
@@ -0,0 +1,133 @@
+
+The available parameters are explained in detail in the
+:meth:`~obspy.clients.syngine.client.Client.get_waveforms()` method and on
+the `Syngine <http://ds.iris.edu/ds/products/syngine/>`_ website.
```
http->https

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+    The ObsPy Development Team (devs@obspy.org)
+:license:
+    GNU Lesser General Public License, Version 3
+    (http://www.gnu.org/copyleft/lesser.html)
```
http->https

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        This method is strongly tied to the actual implementation on the
+        server side. The default values and all the exception handling are
+        deferred to the service. Please see `the Syngine documentation
+        <http://ds.iris.edu/ds/products/syngine/>`_ for more details and the
```
http->https

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        This method is strongly tied to the actual implementation on the
+        server side. The default values and all the exception handling are
+        deferred to the service. Please see the `Syngine documentation
+        <http://ds.iris.edu/ds/products/syngine/>`_ for more details and the
```
http->https

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        self._base_url = base_url
+
+    def _get_url(self, path):
+        return "/".join([self._base_url.rstrip("/"), path])
```
The `rstrip` is already done in `__init__`.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                value = t(value)
+                # String arguments are stripped and empty strings are not
+                # allowed.
+                if t is str:
```
`t` is never `str`; it's sometimes `native_str`, which is slightly different...

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                params[key] = t(value)
+
+        # These can be absolute times, relative times or phase relative times.
+        # jo
```
jo?

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            elif isinstance(value, (int, float)):
+                value = float(value)
+            # If a string like object, attempt to parse it to a datetime
+            # object, otherwise assume its a phase-relative time and let the
```
s/its/it's/

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                value = float(value)
+            # If a string like object, attempt to parse it to a datetime
+            # object, otherwise assume its a phase-relative time and let the
+            # Syngine service deal with the erorr handling.
```
s/erorr/error/

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                try:
+                    value = obspy.UTCDateTime(value)
+                except:
+                    value = (value)
```
Redundant parentheses (though maybe you could just `pass` here.)

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+    def __read_to_stream(self, r):
+        with io.BytesIO(r.content) as buf:
+            # First try to read the file in a normal way, otherwise assume
+            # its a saczip file.
```
s/its/it's/

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                st = obspy.read(buf)
+            except:
+                st = obspy.Stream()
+                zip_obj = zipfile.ZipFile(io.BytesIO(r.content))
```
Can you not just `buf.seek(0)` and reuse the existing `BytesIO` object?

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                    # Skip the log file.
+                    if name.lower() == "syngine.log":
+                        continue
+                    st += obspy.read(io.BytesIO(zip_obj.read(name)))
```
I don't understand why the top-level `BytesIO` object gets put in a `with` context but this one doesn't.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                    if name.lower() == "syngine.log":
+                        continue
+                    st += obspy.read(io.BytesIO(zip_obj.read(name)))
+                return st
```
This `return` is redundant.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        :param model: Specify the model.
+        :type model: str
+        :param network: Specify a network code combined with ``station`` to
+            identify receiver coordinates of a operating station.
```
s/a /an /

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            identify receiver coordinates of a operating station.
+        :type network: str
+        :param station: Specify a network code combined with ``network`` to
+            identify receiver coordinates of a operating station.
```
s/a /an /

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        :param network: Specify a network code combined with ``station`` to
+            identify receiver coordinates of a operating station.
+        :type network: str
+        :param station: Specify a network code combined with ``network`` to
```
s/network code/station code/

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        :type receiverlatitude: float
+        :param receiverlongitude: Specify the receiver longitude in degrees.
+        :type receiverlongitude: float
+        :param networkcode: Specify the network code for synthetic. Optional
```
synthetic what? Maybe you mean synthetic*s*.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            If the value is recognized as a date and time, it is interpreted
+            as an absolute time. If the value is in the form
+            ``phase[+-]offset`` it is interpreted as a phase-relative time,
+            for example ``P-10`` (meaning Pwave arrival time minus 10
```
s/Pwave/P wave/

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            for example ``P-10`` (meaning Pwave arrival time minus 10
+            seconds). If the value is a numerical value it is interpreted as an
+            offset, in seconds, from the ``origintime``.
+        :type starttime: :class:`~obspy.core.utcdatetime.UTCDateTime`
```
Not the only type accepted.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            If the value is recognized as a date and time, it is interpreted
+            as an absolute time. If the value is in the form
+            ``phase[+-]offset`` it is interpreted as a phase-relative time,
+            for example ``P-10`` (meaning Pwave arrival time minus 10
```
s/Pwave/P wave/

Also, since this is the `endtime`, maybe change the example to `P+10`.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            ``phase[+-]offset`` it is interpreted as a phase-relative time,
+            for example ``P-10`` (meaning Pwave arrival time minus 10
+            seconds). If the value is a numerical value it is interpreted as an
+            offset, in seconds, from the ``starttime``.
```
But the bullet list above says it's from origin time.

krischer replied:
Yes. Relative start times are relative to the origin time and relative end times relative to the start time. A bit confusing at first but actually nicer to use.

megies replied:
> Yes. Relative start times are relative to the origin time and relative end times relative to the start time. A bit confusing at first but actually nicer to use.

Still, @QuLogic got a point there, the docstring should be changed `* an offset from origin time in seconds` is not correct for parameter `endtime`, it's an offset from `starttime`.

krischer replied:
Ah I see. I did not look at the bulletpoints. Fixed.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            for example ``P-10`` (meaning Pwave arrival time minus 10
+            seconds). If the value is a numerical value it is interpreted as an
+            offset, in seconds, from the ``starttime``.
+        :type endtime: :class:`~obspy.core.utcdatetime.UTCDateTime`
```
Not the only type accepted.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        :param format: Specify output file to be either miniSEED or a ZIP
+            archive of SAC files, either ``miniseed`` or ``saczip``.
+        :type format: str
+        :param filename: Will download directly to the given file. If given,
```
s/given/specified/

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+        >>> from obspy.clients.syngine import Client
+        >>> c = Client()
+        >>> bulk = [
+        ...     {"netcode": "IU" ,"stacode": "ANMO"},  # net/sta codes
```
s/ ,/, /

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            If the value is recognized as a date and time, it is interpreted
+            as an absolute time. If the value is in the form
+            ``phase[+-]offset`` it is interpreted as a phase-relative time,
+            for example ``P-10`` (meaning Pwave arrival time minus 10
```
s/Pwave/P wave/

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            If the value is recognized as a date and time, it is interpreted
+            as an absolute time. If the value is in the form
+            ``phase[+-]offset`` it is interpreted as a phase-relative time,
+            for example ``P-10`` (meaning Pwave arrival time minus 10
```
s/Pwave/P wave/ + same comment about example time from the other method.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            for example ``P-10`` (meaning Pwave arrival time minus 10
+            seconds). If the value is a numerical value it is interpreted as an
+            offset, in seconds, from the ``origintime``.
+        :type starttime: :class:`~obspy.core.utcdatetime.UTCDateTime`
```
Not the only type accepted here.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            for example ``P-10`` (meaning Pwave arrival time minus 10
+            seconds). If the value is a numerical value it is interpreted as an
+            offset, in seconds, from the ``starttime``.
+        :type endtime: :class:`~obspy.core.utcdatetime.UTCDateTime`
```
Not the only type accepted here.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+            ``phase[+-]offset`` it is interpreted as a phase-relative time,
+            for example ``P-10`` (meaning Pwave arrival time minus 10
+            seconds). If the value is a numerical value it is interpreted as an
+            offset, in seconds, from the ``starttime``.
```
Again, the bullet list above says origin time.

QuLogic commented on obspy/clients/syngine/client.py:
```diff
@@ -0,0 +1,602 @@
+                                "latitude and longitude if either is given." %
+                                str(item))
+                        bulk_item = "{latitude} {longitude}".format(**item)
+                        for _i in ("netcode", "stacode", "loccode"):
```
Why is it `netcode`/`stacode`/`loccode` for bulk requests when the parameter names for the individual requests are called `network`/`station`/`location`?

krischer replied:
Yes it is a bit confusing. `networkcode` and `stationcode` can be used to associate network and station codes for events with custom receivers. `network` and `station` will actually be used to lookup the coordinates of real existing stations and use those for the coordinate lookup.

You are right for the second part. I'll change the bulk request to also take `networkcode`, `stationcode`, and `locationcode` parameters.

QuLogic commented on obspy/clients/syngine/tests/test_client.py:
```diff
@@ -0,0 +1,489 @@
+from obspy.core.util.base import NamedTemporaryFile
+from obspy.clients.syngine import Client
+from obspy.clients.base import DEFAULT_TESTING_USER_AGENT, ClientHTTPException
+
```
Missing newline.

QuLogic commented on obspy/clients/syngine/tests/test_client.py:
```diff
@@ -0,0 +1,489 @@
+        """
+        info = self.c.get_model_info("test")
+
+        self.assertTrue(isinstance(info, obspy.core.AttribDict))
```
`self.assertTrue(isinstance(` -> `self.assertIsInstance(`.

krischer replied:
Nice. I did not know that this is a thing!

QuLogic commented on obspy/clients/syngine/tests/test_client.py:
```diff
@@ -0,0 +1,489 @@
+        models = self.c.get_available_models()
+        self.assertTrue(isinstance(models, dict))
+        keys = list(models.keys())
+        self.assertTrue(len(keys) > 3)
```
`self.assertGreater`

QuLogic commented on obspy/clients/syngine/tests/test_client.py:
```diff
@@ -0,0 +1,489 @@
+    def test_get_available_models(self):
+        models = self.c.get_available_models()
+        self.assertTrue(isinstance(models, dict))
+        keys = list(models.keys())
```
I'm not sure how `future` handles it, but I don't think you actually need to convert this into a list.

However, you don't need to get the keys: `len(list(models.keys()))` -> `len(models.keys())` -> `len(models)` and `'name' in list(models.keys())` -> `'name' in models.keys()` -> `'name' in models`.

QuLogic commented on obspy/clients/syngine/tests/test_client.py:
```diff
@@ -0,0 +1,489 @@
+
+    def test_error_handling_arguments(self):
+        # Floating points value
+        self.assertRaises(ValueError, self.c.get_waveforms, model="test",
```
I would use this as a context manager instead to preserve the call syntax.

QuLogic commented on obspy/clients/syngine/tests/test_client.py:
```diff
@@ -0,0 +1,489 @@
+
+    def test_error_handling(self):
+        """
+        Tests the error handling. The clients just passes on most things to
```
s/passes/pass/

QuLogic commented on obspy/clients/syngine/tests/test_client.py:
```diff
@@ -0,0 +1,489 @@
+    def test_error_handling(self):
+        """
+        Tests the error handling. The clients just passes on most things to
+        syngine and relies on the service for the error detection.
```
s/relies/rely/

QuLogic commented:
Alright, _now_ I'm done review.

krischer commented:
Thanks a bunch @QuLogic 

I implemented all review points and rebased on top of master to resolve a couple of conflicts. Good to merge from my point of view.

megies commented:
One tiny thing still to be fixed: https://github.com/obspy/obspy/pull/1227#discussion_r51588308 (https://github.com/obspy/obspy/pull/1227/files#diff-13f36e73dd1c2a80ee5d82baefe7dd58R302)

QuLogic commented on obspy/clients/base.py:
```diff
@@ -50,31 +50,198 @@ def get_stations(self, network, station, location, channel, starttime,
+        """
+        Download the URL with GET or POST and the chosen parameters.
+
+        Will only return if it manages to download with HTTP code 200,
```
This is still incorrect, since it depends on `filename`, not HTTP status code.

QuLogic commented:
Just one more thing on the bulk request. IIUC, in the individual request, you can use either `network`/`station`/`location` or `latitude`/`longitude` to select a location, and `networkcode`/`stationcode`/`locationcode` to set the identifier for the output. In the bulk request, the `netcode`/`stacode`/`loccode` were renamed to `networkcode`/`stationcode`/`locationcode` but those names seem to be used for _both_ selecting the location _and_ setting the identifier.

krischer commented:
Yea you are right - that is slightly confusing. I'll change it to either allow `network,station` or `latitude,longitude[,networkcode,stationcode,locationcode]`. It should then be clear what is happening.

krischer commented:
All done. I hope its good now.

QuLogic commented:
Looks good. One last question; does it support specifying `network`/`station` _and_ an overriding `networkcode`/`stationcode`? I don't think any of the examples/tests do so, but I don't know if the service supports it either.

krischer commented:
> Looks good. One last question; does it support specifying network/station and an overriding networkcode/stationcode? I don't think any of the examples/tests do so, but I don't know if the service supports it either.

The service does not support it and the ObsPy client just ignores the `networkcode`/`stationcode` keys in that case and does not include them in the POST request.

QuLogic commented:
Alright then...

dbgraybeal commented:
Hi @krischer, when this was written, did the Syngine Green's function database exist? I prefer using this client over building my own URLs, and get_waveforms() seems to handle both synthetic seismograms and Green's Functions perfectly with the addition of the two required parameters, greensfunction and sourcedistanceindegrees. 

If I wrote a PR to add Green's function downloading, would it be preferable to include it in get_waveforms() or to duplicate some code and create a get_greensfunctions() function? It seems to me that both seismograms and Green's functions are "waveforms," since I expected the functionality to already be in there and just not highlighted in the documentation.
